Utilities/Mergers/src/Merger.cxx

```diff
@@ -0,0 +1,284 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file Merger.cxx
+/// \brief Implementation of O2 Mergers, v1.0
+///
+/// \author Piotr Konopka, piotr.jan.konopka@cern.ch
+
+#include "Mergers/Merger.h"
+#include "Mergers/MergerBuilder.h"
+
+#include <Framework/CompletionPolicyHelpers.h>
+#include <Framework/TimesliceIndex.h>
+#include <Framework/CallbackService.h>
+
+#include <TObjArray.h>
+#include <TH1.h>
+#include <TH2.h>
+#include <TH3.h>
+#include <THn.h>
+#include <TTree.h>
+#include <THnSparse.h>
+
+using namespace o2::framework;
+using namespace std::chrono;
+
+namespace o2
+{
+namespace mergers
+{
+
+Merger::Merger(MergerConfig config, header::DataHeader::SubSpecificationType subSpec)
+  : mConfig(config),
+    mSubSpec(subSpec),
+    mCache(config.ownershipMode.value == OwnershipMode::Full)
+{
+}
+
+Merger::~Merger()
+{
+}
+
+void Merger::init(framework::InitContext& ictx)
+{
+  if (mConfig.publicationDecision.value == PublicationDecision::EachNSeconds) {
+    // Register a device callback which creates timeslice in the TimesliceIndex
+    // each N seconds, so it can serve as timer input.
+    ictx.services().get<CallbackService>().set(CallbackService::Id::ClockTick, prepareTimerCallback(ictx));
+  }
+}
+
+void Merger::run(framework::ProcessingContext& ctx)
+{
+  if (mCache.uninitialized()) {
+    mCache.init(ctx.inputs());
+  }
+
+  mCache.cacheInputRecord(ctx.inputs());
+
```

Comment:
it looks like the order of calling the mCache methods to initialize the object matters here - is that not a bit dangerous?

Reply:
I am not 100% sure what do you mean. Do you suggest to move
```
  if (mCache.uninitialized()) {
    mCache.init(ctx.inputs());
  }
```
inside the cacheInputRecord(), so it is the Cache responsibility to make sure it is initialized?

Reply:
something like that - an object always has to be in a valid state - don't split "atomic" operations into two calls.